ffi/src/handle.rs

```diff
@@ -1,235 +1,616 @@
-///
-/// * The `Target` type is `Sync`
-///
-/// * The handle is (correctly) `Send`
-pub trait ArcHandle: Sized {
-    /// The target type this handle represents.
-    type Target: ?Sized + Send + Sync;
+pub trait HandleDescriptor {
+    /// The true type of the handle's underlying object
+    type Target: ?Sized + Send;
 
-    /// Converts the target Arc into a (leaked) "handle" that can cross the FFI boundary. The Arc
-    /// refcount does not change. The handle remains valid until passed to [Self::drop_handle].
-    fn into_handle(target: Arc<Self::Target>) -> *const Self {
-        Box::into_raw(Box::new(target)) as _
-    }
+    /// If `True`, the handle owns the underlying object (`Box`-like); otherwise, the handle owns a
+    /// shared reference to the underlying object (`Arc`-like).
+    type Mutable: Boolean;
+
+    /// `True` iff the handle's `Target` is [`Sized`].
+    type Sized: Boolean;
+}
+
+mod private {
+
+    use std::mem::ManuallyDrop;
+    use std::ptr::NonNull;
+    use std::sync::Arc;
 
-    /// Clones an Arc from an FFI handle for kernel use, without dropping the handle. The Arc
-    /// refcount increases by one.
+    use super::*;
+
+    /// Represents an object that crosses the FFI boundary and which outlives the scope that created
+    /// it. It can be passed freely between rust code and external code.
     ///
-    /// # Safety
+    /// # Validity
     ///
-    /// Caller of this method asserts that `handle` satisfies all of the following conditions:
+    /// A `Handle` is _valid_ if all of the following hold:
     ///
-    /// * Obtained by calling [into_handle]
-    /// * Never cast to any other type nor dereferenced
-    /// * Not previously passed to [drop_handle]
-    unsafe fn clone_as_arc(handle: *const Self) -> Arc<Self::Target> {
-        let ptr = handle as *mut Arc<Self::Target>;
-        let arc = unsafe { &*ptr };
-        arc.clone()
+    /// * It was created by a call to [`Handle::from`]
+    /// * Not yet dropped by a call to [`Handle::drop_handle`]
+    /// * Not yet consumed by a call to [`Handle::into_inner`]
+    ///
+    /// Additionally, in keeping with the [`Send`] contract, multi-threaded external code must
+    /// enforce mutual exclusion -- no mutable handle should ever be passed to more than one kernel
+    /// API call at a time. If thread races are possible, the handle should be protected with a
+    /// mutex. Due to Rust [reference
+    /// rules](https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html#the-rules-of-references),
+    /// this requirement applies even for API calls that appear to be read-only (because Rust code
+    /// always receives the handle as mutable).
+    ///
+    /// NOTE: Because the underlying type is always [`Sync`], multi-threaded external code can
+    /// freely access shared (non-mutable) handles.
+    ///
+    /// cbindgen:transparent-typedef
+    #[repr(transparent)]
+    pub struct Handle<H: HandleDescriptor> {
+        ptr: NonNull<H>,
```

Comment:
TODO: Make the shared handles produce `const T*` on C/C++ side.

Reply:
Update: `NonNull<T>` always produces `T*` (mutable) pointers. Nothing we can do about that, unfortunately.